custom_metrics/structured-data.js

```diff
@@ -0,0 +1,271 @@
+            rdfa_vocabs: [],
+            rdfa_prefixes: [],
+            rdfa_typeofs: [],
+            microformats2_types: [],
+            microformats_classic_types: [],
+            meta_tags: [],
+            link_tags: [],
+            dublin_core: [],
+            twitter: [],
+            facebook: [],
+            opengraph: [],
+          };
+
+          // JSON-LD
+          target.jsonld_scripts = [
+            ...d.querySelectorAll('script[type="application/ld+json"i]'),
+          ].map((script) => script.innerHTML);
+          target.present.json_ld = target.jsonld_scripts.length > 0;
+
+          // Microdata, RDFa & Microformats
+          target.raw_html = d.documentElement.outerHTML;
+
+          // Microdata
+          target.present.microdata =
+            d.querySelectorAll("[itemscope],[itemtype],[itemprop]").length > 0;
+          target.microdata_itemtypes = [
+            ...d.querySelectorAll("[itemtype]"),
+          ].map((element) => element.getAttribute("itemtype"));
+
+          // RDFa
+          target.present.rdfa =
+            d.querySelectorAll(
+              "[vocab],[typeof],[property],[resource],[prefix]"
+            ).length > 0;
+          target.rdfa_vocabs = getAttributeValues(d, "vocab");
+          target.rdfa_prefixes = getAttributeValues(d, "prefix");
+          target.rdfa_typeofs = getAttributeValues(d, "typeof");
+
+          // microformats2
+          [
+            "h-adr",
+            "h-card",
+            "h-entry",
+            "h-event",
+            "h-feed",
+            "h-geo",
+            "h-item",
+            "h-listing",
+            "h-product",
+            "h-recipe",
+            "h-resume",
+            "h-review",
+            "h-review-aggregate",
+          ].forEach((name) => {
+            let items = d.querySelectorAll('[class~="' + name + '" i]');
+            if (items.length > 0) {
+              target.present.microformats2 = true;
+              target.microformats2_types.push({
+                name: name,
+                count: items.length,
+              });
+            }
+          });
+
+          // Classic Microformats
+          [
+            "hAtom",
+            "hCalendar",
+            "hCard",
+            "hListing",
+            "hMedia",
+            "hProduct",
+            "hRecipe",
+            "hResume",
+            "hReview",
+            "hReview-aggregate",
+            "adr",
+            "geo",
+          ].forEach((name) => {
+            let items = d.querySelectorAll('[class~="' + name + '" i]');
+            if (items.length > 0) {
+              target.present.microformats_classic = true;
+              target.microformats_classic_types.push({
+                name: name,
+                count: items.length,
+              });
+            }
+          });
+
+          // Dublin Core, Twitter, Facebook & OpenGraph
+          target.meta_tags = dumpTagAttributes([...d.querySelectorAll("meta")]);
```

Comment:
We do definitely want to have the `content` values for these tags, so I think we should still do our own thing separate to what's in `almanac.js`, but we've restricted our storage to just the following attributes now: `["rel", "href", "name", "property", "content"]`.

Reply:
I raised the question because it's not immediately clear how all meta/link elements are related to structured data. There would need to be an additional post-processing step to analyze the attributes and determine which ones are relevant to the chapter and aggregate/summarize them in a useful way. Usually that's done at the same time as the custom metric processing to simplify the SQL workload.

Is your intention to emit more data than you need so that you can analyze it later? If so this would be similar to my feedback on the raw HTML, that we should process the data now.